Feat/notarization tutorial (#682)
iota-wiki/iota-wiki: content/tutorials/single-page-tutorials/proof-inclusion-of-a-block.md

# Tokenless Data Notarization

> This tutorial shows how you can prove that a block was part of the Tangle, even if this part of the Tangle was already pruned from network nodes.

Reviewer: ```suggestion
This tutorial shows how you can prove that a block was part of the Tangle, even if it was already pruned from network nodes.
```

## User Story

A **Prover** wants to prove to a **Verifier** that a dataset or file was not altered by notarizing it using the public and permissionless Tangle. 

> A good example could be the government (**Verifier**) obligating companies (**Prover**) to write their daily CO2 emissions to the Tangle to create immutable logs. The government has no interest in providing storage for all the companies but wants to receive verifiable, immutable data in the case of an audit. Therefore, all the data remains with the respective company until an audit occurs.

Reviewer: ```suggestion
A good example could be the government (**Verifier**) obligating companies (**Prover**) to write their daily CO₂ emissions to the Tangle to create immutable logs. The government has no interest in providing storage for all the companies but wants to receive verifiable, immutable data in the case of an audit. Therefore, all the data remains with the respective company until an audit occurs.
```

### Architecture

> In this use case, the party writing to the Tangle (**Prover**) is the same party reading from it. This allows reading a block by its block Id together with proof. All a **Verifier** needs to verify this proof is the chain of milestones back to the milestone that referenced this exact block. Every milestone directly references the previous milestone, which leads to a unique and verifiable chain of milestones in the Tangle. Even if the **Verifier's** notarization plugin didn't record the milestone history already, it could still download the milestone chain from another source and verify its correctness before using it to notarize verifications.

Reviewer: > All a **Verifier** needs to verify this proof

I think some words are missing there.

> Since the pattern assumes trusted access to the network, we recommend that both the **Prover** and the **Verifier** operate their own network node and run the [INX Notarization/PoI Plugin](https://github.com/iotaledger/inx-poi). It is also noteworthy that this pattern does not rely on a Permanode storing the entire Tangle history, drastically reducing the node operator's storage requirements.

Reviewer: Please explain how trusted access is related to the INX Notarization/PoI plugin, as a reader I am missing a connection here.

> Since the pattern assumes trusted access to the network, we recommend that both the **Prover** and the **Verifier** operate their own network node and run the [INX Notarization/PoI Plugin](https://github.com/iotaledger/inx-poi). It is also noteworthy that this pattern does not rely on a Permanode storing the entire Tangle history, drastically reducing the node operator's storage requirements.

Reviewer: ```suggestion
Since the pattern assumes trusted access to the network, we recommend that both the **Prover** and the **Verifier** operate their own network node and run the [INX Notarization/PoI Plugin](https://github.com/iotaledger/inx-poi). As this pattern does not rely on a Permanode storing the entire Tangle history, it consumes much less storage space.
```

- The node should run the [INX-POI](https://github.com/iotaledger/inx-poi). The node(s) will start the INX-POI plugin by default.

### Development Environment and Libraries

- [Node.js](https://nodejs.org/en/).

## Set Up

### Prepare Your Development Environment

1. Create a new folder for the tutorial in a location of your choice, for example:

```bash
mkdir notarization-tutorial
cd notarization-tutorial
```

2.  You should run the Node.js initializer to configure the project.

```bash
npm init --yes
```

3. Add [`@iota/iota.js`](https://www.npmjs.com/package/@iota/iota.js) as a dependency by running the following command:

```bash
npm install @iota/iota.js
```

5. Install dependencies by running the following command:

```bash
npm install
```

## Create Your Scripts

### Network Configuration Script

Both the script that will [create](#create-notarization-script) and [verify](#verify-notarization-script) the notarization will need to share a network.

> You should create a new file named `networkConfig.js` and add the following code:

Reviewer: ```suggestion

Create a new file named `networkConfig.js` and add the following code:
```

      await sleep(1000);
      const blockMetadata = await client.blockMetadata(blockId);
      // If a block was referenced by a milestone, the node will return its metadata with the key 'referencedByMilestoneIndex', otherwise the key won't be there
      if ('referencedByMilestoneIndex' in blockMetadata) {
        console.log(
          `Try ${i}: Block was referenced by milestone #${blockMetadata.referencedByMilestoneIndex}`,
          '\n',
        );
        // Call "create" endpoint of PoI plugin with blockId and return the result
        const poiPluginUrl = `${nodeURL}/api/poi/v1/create/${blockId}`;
        const response = await fetch(poiPluginUrl);
        const result = await response.json();
        return result;
      } else {
        console.log(`Try ${i}: Block was not yet referenced by a milestone`);
      }
    }
    console.log(`Block was not referenced by a milestone after ${i} seconds.`);
    return false;
  } catch (error) {
    console.log(error);
  }
}
function sleep(ms) {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}
run().catch((err) => console.error(err));
```

### Verify Notarization Script

> Since you want to verify the notarization, you will also need to create a new file called `verify-notarization.js`. 

Reviewer: ```suggestion
Since you want to verify the notarization, you also need to create a new file called `verify-notarization.js`. 
```

The `verify-notorization.js` will:

1. Read the block with the notarization from your local folder.
2. Verify the block with the notarization.

> You should create a new file called `verify-notarization.js` and add the following code:

Reviewer: At this point that file is already created.

Author: No it hasn't. The description follows the same order as the previously created network script and create notarization script.

    console.log(filePath, '\n');
    // Generate blockId from block content and log explorer link
    // The blockId is defined as the BLAKE2b-256 hash of the entire serialized block
    const blockId = TransactionHelper.calculateBlockId(notarizedBlock.block);
    console.log(consoleColor, 'Notarized block:');
    console.log(explorerURL+"block/"+blockId, '\n');
    // Verify provided notarization/proof of inclusion for block
    const validity = await verifyNotarization(nodeURL, notarizedBlock);
    console.log(consoleColor, 'Validity of provided notarization:');
    console.log(validity, '\n');
}
async function verifyNotarization(nodeURL, notarizedBlock) {
    // Call "validate" endpoint of PoI plugin with notarized block and return boolean answer
    const poiPluginUrl = `${nodeURL}/api/poi/v1/validate`;
    const response = await fetch(poiPluginUrl, {
        method: 'POST', 
        body: JSON.stringify(notarizedBlock),
        headers: { 'Content-Type': 'application/json' }
    })
    const result = await response.json();
    return result.valid;
}
run().catch((err) => console.error(err));
```

## Run Your Scripts

> Once you have [created your scripts](#create-your-scripts), you can execute the two created files in the order. 

Reviewer: ```suggestion
Once you have [created your scripts](#create-your-scripts), you can execute the two created files in order. 
```

Naturally, you will need to run `create-notarization` before you can run `verify-notarization` and check the log output to follow along.

### Create the Notarization

You can create the notarization in the Tangle by running the following command:

```bash
node create-notarization.js
```

> Your console output should look something like this:

Reviewer: ```suggestion
Your console output should look similar to this:
```

```plaintext
Attached block:
http://localhost:8082/dashboard/explorer/block/<id_of_your_notarized_block>
Wait for milestone confirmation to get notarized block:
Try 1: Block was not yet referenced by a milestone
Try 2: Block was not yet referenced by a milestone
Try 3: Block was referenced by milestone <number_of_milestone_referencing_your_block>
Block successfully notarized and stored at:
./notarized-block.json
Notarized block can now be handed over to the verifier
```

### Verify Notarization

> 
> 

Reviewer: ```suggestion

```

You can verify the notarization by running the following command:

```bash
node verify-notarization.js
```

> Your console output should look something like this:

Reviewer: ```suggestion
Your console output should look similar to this:
```

### Architecture

In this use-case pattern, the party writing to the Tangle (**Prover**) is the same party reading from it, which allows reading a block by its block Id together with a proof. All a **Verifier** needs to verify this proof, is the chain of milestones back to the milestone that referenced this exact block. Every milestone directly references the previous milestone, which leads to a unique and verifiable chain of milestones in the Tangle. This means even if the **Verifier's** notarization plugin didn't record the milestone history already, it could still download the milestone chain from another source and verify its correctness, before using it for further notarization verifications.

Since the pattern assumes trusted access to the network, it is strongly recommended that both, the **Prover** and the **Verifier** operate their own network node and run the [INX Notarization/PoI Plugin](https://github.com/iotaledger/inx-poi). It is also noteworthy that this pattern does not rely on a Permanode storing the full Tangle history, which drastically reduces the storage requirements on the node operator side.

![PoI-Architecture](proof-inclusion-of-a-block-architecture.png)

> ### Prerequisites

Reviewer 2: ```suggestion
## Prerequisites
```

- Network

  - Running [Hornet Private Tangle](https://github.com/iotaledger/hornet/tree/develop/private_tangle) (Branch **develop** for the latest [Stardust](https://blog.shimmer.network/stardust-upgrade-in-a-nutshell/) features)
  - The node(s) will start with the necessary [INX-POI](https://github.com/iotaledger/inx-poi) plugin by default

- Development Environment and Libraries
  - Installed [Node.js](https://nodejs.org/en/)

> ### Set Up
> 
> #### Prepare development environment
> 
> Create a new folder for the tutorial in a location of your choice.
> 
> ```bash
> mkdir notarization-tutorial
> cd notarization-tutorial
> ```
> 
> Now we need to run the Node.js initializer to configure the project.
> 
> ```bash
> npm init --yes
> ```
> 
> Add the following dependencies to the `package.json` file.
> 
> ```yaml
> {
>   'name': 'proof-of-inclusion',
>   'version': '1.0.0',
>   'description': 'Proof of Inclusion to the Tangle',
>   'main': 'create-notarization.js',
>   'dependencies':
>     {
>       '@iota/iota.js': 'next'
>     },
>   'scripts': { 'test': 'echo "Error: no test specified" && exit 1' },
>   'author': '',
>   'license': 'Apache-2.0',
> }
> ```
> 
> Install dependencies.
> 
> ```bash
> npm install
> ```

Reviewer 2: ````suggestion
## Set Up

### Prepare Your Development Environment

1. Create a new folder for the tutorial in a location of your choice, for example:

```bash
mkdir notarization-tutorial
cd notarization-tutorial
```

2.  You should run the Node.js initializer to configure the project.

```bash
npm init --yes
```

3. Replace the content of the `package.json` with the following:

```json
{
  'name': 'proof-of-inclusion',
  'version': '1.0.0',
  'description': 'Proof of Inclusion to the Tangle',
  'main': 'create-notarization.js',
  'dependencies':
    {
      '@iota/iota.js': 'next'
    },
  'scripts': { 'test': 'echo "Error: no test specified" && exit 1' },
  'author': '',
  'license': 'Apache-2.0',
}
```

4. Install dependencies by running the following command:

```bash
npm install
```
````

> ### Set Up
> 
> #### Prepare development environment
> 
> Create a new folder for the tutorial in a location of your choice.
> 
> ```bash
> mkdir notarization-tutorial
> cd notarization-tutorial
> ```
> 
> Now we need to run the Node.js initializer to configure the project.
> 
> ```bash
> npm init --yes
> ```
> 
> Add the following dependencies to the `package.json` file.
> 
> ```yaml
> {
>   'name': 'proof-of-inclusion',
>   'version': '1.0.0',
>   'description': 'Proof of Inclusion to the Tangle',
>   'main': 'create-notarization.js',
>   'dependencies':
>     {
>       '@iota/iota.js': 'next'
>     },
>   'scripts': { 'test': 'echo "Error: no test specified" && exit 1' },
>   'author': '',
>   'license': 'Apache-2.0',
> }
> ```
> 
> Install dependencies.
> 
> ```bash
> npm install
> ```

Reviewer 2: ````suggestion
## Set Up

### Prepare Your Development Environment

1. Create a new folder for the tutorial in a location of your choice, for example:

```bash
mkdir notarization-tutorial
cd notarization-tutorial
```

2.  You should run the Node.js initializer to configure the project.

```bash
npm init --yes
```

3. Replace the content of the `package.json` with the following:

```json
{
  'name': 'proof-of-inclusion',
  'version': '1.0.0',
  'description': 'Proof of Inclusion to the Tangle',
  'main': 'create-notarization.js',
  'dependencies':
    {
      '@iota/iota.js': 'next'
    },
  'scripts': { 'test': 'echo "Error: no test specified" && exit 1' },
  'author': '',
  'license': 'Apache-2.0',
}
```

4. Install dependencies by running the following command:

```bash
npm install
```
````

    let i = 0;
    while (i < 10) {
      i++;
      // Function waits for a certain time between iterations
      await sleep(1000);

      const blockMetadata = await client.blockMetadata(blockId);

      // If a block was referenced by a milestone the node will return its metadata with the key 'referencedByMilestoneIndex', otherwise the key won't be there
      if ('referencedByMilestoneIndex' in blockMetadata) {
        console.log(
          `Try ${i}: Block was referenced by milestone #${blockMetadata.referencedByMilestoneIndex}`,
          '\n',
        );

        // Call "create" endpoint of PoI plugin with blockId and return the result
        const poiPluginUrl = `${nodeURL}/api/poi/v1/create/${blockId}`;
        const response = await fetch(poiPluginUrl);
        const result = await response.json();

        return result;
      } else {
        console.log(`Try ${i}: Block was not yet referenced by a milestone`);
      }
    }
    console.log(`Block was not referenced by a milestone after ${i} seconds.`);
    return false;
  } catch (error) {
    console.log(error);
  }
}

function sleep(ms) {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });

> }
> 
> run().catch((err) => console.error(err));
> ```
> 
> #### Prepare file for Notarization verification
> 
> Create new file `verify-notarization.js` and add the following code.
> 
> ```javascript
> 
> const { TransactionHelper } = require('@iota/iota.js');
> const fs = require('fs');
> const fetch = require('node-fetch')
> 
> // Network configuration
> const { networkConfig } = require("./networkConfig.js");
> const nodeURL = networkConfig.node;
> const explorerURL = networkConfig.explorer;
> 
> 
> // For the sake of this tutorial, some console output will be printed in a different color for better readability
> const consoleColor = '\x1b[36m%s\x1b[0m';
> 
> 
> async function run() {
>     // Read and parse notarized block from file path
>     const filePath = './notarized-block.json';
>     const file = fs.readFileSync(filePath);
>     const notarizedBlock = JSON.parse(file);
>     console.log(consoleColor, 'Successfully imported notarized block from path:');
>     console.log(filePath, '\n');
> 
>     // Generate blockId from block content and log explorer link
>     // The blockId is defined as the BLAKE2b-256 hash of the entire serialized block
>     const blockId = TransactionHelper.calculateBlockId(notarizedBlock.block);
>     console.log(consoleColor, 'Notarized block:');
>     console.log(explorerURL+"block/"+blockId, '\n');
> 
>     // Verify provided notarization/proof of inclusion for block
>     const validity = await verifyNotarization(nodeURL, notarizedBlock);
>     console.log(consoleColor, 'Validity of provided notarization:');
>     console.log(validity, '\n');
> }
> 
> async function verifyNotarization(nodeURL, notarizedBlock) {
>     // Call "validate" endpoint of PoI plugin with notarized block and return boolean answer
>     const poiPluginUrl = `${nodeURL}/api/poi/v1/validate`;
>     const response = await fetch(poiPluginUrl, {
>         method: 'POST', 
>         body: JSON.stringify(notarizedBlock),
>         headers: { 'Content-Type': 'application/json' }
>     })
>     const result = await response.json();
> 
>     return result.valid;
> }
> 
> run().catch((err) => console.error(err));
> ```

Reviewer 2: ````suggestion
## Create Your Scripts

### Network Configuration Script

Both the script that will [create](#create-notarization-script) and [verify](#verify-notarization-script) the notarization will need to share a network.
You should create a new file named `networkConfig.js` and add the following code:

```javascript
var networkConfig = {};
// Private Tangle Network
networkConfig.node = 'http://localhost:14265';
networkConfig.faucet = 'http://localhost:8091';
networkConfig.explorer = 'http://localhost:8082/dashboard/explorer/';
module.exports = { networkConfig };
```

### Create Notarization Script

The first script you will need to create is `create-notorization.js`.
The `create-notorization.js` script will:

1. Set up the client and define block content.
2. Attach the block to the Tangle.
3. Wait for the block to be confirmed by a milestone.
4. Read the block from Tangle, together with the notarization.
5. Store the block with its notarization in a JSON file within your local folder.

You should create a new file called `create-notarization.js` and add the following code:

```javascript
const {
  SingleNodeClient,
  LocalPowProvider,
  sendData,
} = require('@iota/iota.js');
const fs = require('fs');
const fetch = require('node-fetch');
// Network configuration
const { networkConfig } = require('./networkConfig.js');
const nodeURL = networkConfig.node;
const explorerURL = networkConfig.explorer;
// For the sake of this tutorial, some console output will be printed in a different color for better readability
const consoleColor = '\x1b[36m%s\x1b[0m';
async function run() {
  // Setup client and define block content
  const client = new SingleNodeClient(nodeURL, {
    powProvider: new LocalPowProvider(),
  });
  const tag = 'This is my Tag';
  const data = 'This is my data';
  // Attach block to Tangle and log explorer link
  const sendResult = await sendData(client, tag, data);
  const blockId = sendResult.blockId;
  console.log(consoleColor, 'Attached block:');
  console.log(explorerURL + 'block/' + blockId, '\n');
  // Wait for block confirmation by milestone and read it with proof of inclusion from INX plugin
  const result = await getNotarization(client, nodeURL, blockId);
  // Store block with proof of inclusion in local json file
  if (result != false) {
    const filePath = `./notarized-block.json`;
    fs.writeFileSync(filePath, JSON.stringify(result, null, 4));
    console.log(consoleColor, 'Block successfully notarized and stored at:');
    console.log(filePath, '\n');
    console.log(
      consoleColor,
      'The Notarized block can now be handed over to the verifier',
    );
  }
}
// Function that regularly checks for block confirmation and returns proof of inclusion if confirmed after n tries
async function getNotarization(client, nodeURL, blockId) {
  try {
    console.log(
      consoleColor,
      'Wait for milestone confirmation to get notarized block:',
    );
    let i = 0;
    while (i < 10) {
      i++;
      // Function waits for a certain time between iterations
      await sleep(1000);
      const blockMetadata = await client.blockMetadata(blockId);
      // If a block was referenced by a milestone, the node will return its metadata with the key 'referencedByMilestoneIndex', otherwise the key won't be there
      if ('referencedByMilestoneIndex' in blockMetadata) {
        console.log(
          `Try ${i}: Block was referenced by milestone #${blockMetadata.referencedByMilestoneIndex}`,
          '\n',
        );
        // Call "create" endpoint of PoI plugin with blockId and return the result
        const poiPluginUrl = `${nodeURL}/api/poi/v1/create/${blockId}`;
        const response = await fetch(poiPluginUrl);
        const result = await response.json();
        return result;
      } else {
        console.log(`Try ${i}: Block was not yet referenced by a milestone`);
      }
    }
    console.log(`Block was not referenced by a milestone after ${i} seconds.`);
    return false;
  } catch (error) {
    console.log(error);
  }
}
function sleep(ms) {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}
run().catch((err) => console.error(err));
```

### Verify Notarization Script

Since you want to verify the notarization, you will also need to create a new file called `verify-notarization.js`. 

The `verify-notorization.js` you created will do the following:

1. Read the block with the notarization from your local folder.
2. Verify the block with the notarization.

You should create a new file called `verify-notarization.js` and add the following code:

```javascript
const { TransactionHelper } = require('@iota/iota.js');
const fs = require('fs');
const fetch = require('node-fetch')
// Network configuration
const { networkConfig } = require("./networkConfig.js");
const nodeURL = networkConfig.node;
const explorerURL = networkConfig.explorer;
// For the sake of this tutorial, some console output will be printed in a different color for better readability
const consoleColor = '\x1b[36m%s\x1b[0m';
async function run() {
    // Read and parse notarized block from file path
    const filePath = './notarized-block.json';
    const file = fs.readFileSync(filePath);
    const notarizedBlock = JSON.parse(file);
    console.log(consoleColor, 'Successfully imported notarized block from path:');
    console.log(filePath, '\n');
    // Generate blockId from block content and log explorer link
    // The blockId is defined as the BLAKE2b-256 hash of the entire serialized block
    const blockId = TransactionHelper.calculateBlockId(notarizedBlock.block);
    console.log(consoleColor, 'Notarized block:');
    console.log(explorerURL+"block/"+blockId, '\n');
    // Verify provided notarization/proof of inclusion for block
    const validity = await verifyNotarization(nodeURL, notarizedBlock);
    console.log(consoleColor, 'Validity of provided notarization:');
    console.log(validity, '\n');
}
async function verifyNotarization(nodeURL, notarizedBlock) {
    // Call "validate" endpoint of PoI plugin with notarized block and return boolean answer
    const poiPluginUrl = `${nodeURL}/api/poi/v1/validate`;
    const response = await fetch(poiPluginUrl, {
        method: 'POST', 
        body: JSON.stringify(notarizedBlock),
        headers: { 'Content-Type': 'application/json' }
    })
    const result = await response.json();
    return result.valid;
}
run().catch((err) => console.error(err));
```
````

> }
> 
> run().catch((err) => console.error(err));
> ```
> 
> #### Prepare file for Notarization verification
> 
> Create new file `verify-notarization.js` and add the following code.
> 
> ```javascript
> 
> const { TransactionHelper } = require('@iota/iota.js');
> const fs = require('fs');
> const fetch = require('node-fetch')
> 
> // Network configuration
> const { networkConfig } = require("./networkConfig.js");
> const nodeURL = networkConfig.node;
> const explorerURL = networkConfig.explorer;
> 
> 
> // For the sake of this tutorial, some console output will be printed in a different color for better readability
> const consoleColor = '\x1b[36m%s\x1b[0m';
> 
> 
> async function run() {
>     // Read and parse notarized block from file path
>     const filePath = './notarized-block.json';
>     const file = fs.readFileSync(filePath);
>     const notarizedBlock = JSON.parse(file);
>     console.log(consoleColor, 'Successfully imported notarized block from path:');
>     console.log(filePath, '\n');
> 
>     // Generate blockId from block content and log explorer link
>     // The blockId is defined as the BLAKE2b-256 hash of the entire serialized block
>     const blockId = TransactionHelper.calculateBlockId(notarizedBlock.block);
>     console.log(consoleColor, 'Notarized block:');
>     console.log(explorerURL+"block/"+blockId, '\n');
> 
>     // Verify provided notarization/proof of inclusion for block
>     const validity = await verifyNotarization(nodeURL, notarizedBlock);
>     console.log(consoleColor, 'Validity of provided notarization:');
>     console.log(validity, '\n');
> }
> 
> async function verifyNotarization(nodeURL, notarizedBlock) {
>     // Call "validate" endpoint of PoI plugin with notarized block and return boolean answer
>     const poiPluginUrl = `${nodeURL}/api/poi/v1/validate`;
>     const response = await fetch(poiPluginUrl, {
>         method: 'POST', 
>         body: JSON.stringify(notarizedBlock),
>         headers: { 'Content-Type': 'application/json' }
>     })
>     const result = await response.json();
> 
>     return result.valid;
> }
> 
> run().catch((err) => console.error(err));
> ```

Reviewer 2: ````suggestion
## Create Your Scripts

### Network Configuration Script

Both the script that will [create](#create-notarization-script) and [verify](#verify-notarization-script) the notarization will need to share a network.
You should create a new file named `networkConfig.js` and add the following code:

```javascript
var networkConfig = {};
// Private Tangle Network
networkConfig.node = 'http://localhost:14265';
networkConfig.faucet = 'http://localhost:8091';
networkConfig.explorer = 'http://localhost:8082/dashboard/explorer/';
module.exports = { networkConfig };
```

### Create Notarization Script

The first script you will need to create is `create-notorization.js`.
The `create-notorization.js` script will:

1. Set up the client and define block content.
2. Attach the block to the Tangle.
3. Wait for the block to be confirmed by a milestone.
4. Read the block from Tangle, together with the notarization.
5. Store the block with its notarization in a JSON file within your local folder.

You should create a new file called `create-notarization.js` and add the following code:

```javascript
const {
  SingleNodeClient,
  LocalPowProvider,
  sendData,
} = require('@iota/iota.js');
const fs = require('fs');
const fetch = require('node-fetch');
// Network configuration
const { networkConfig } = require('./networkConfig.js');
const nodeURL = networkConfig.node;
const explorerURL = networkConfig.explorer;
// For the sake of this tutorial, some console output will be printed in a different color for better readability
const consoleColor = '\x1b[36m%s\x1b[0m';
async function run() {
  // Setup client and define block content
  const client = new SingleNodeClient(nodeURL, {
    powProvider: new LocalPowProvider(),
  });
  const tag = 'This is my Tag';
  const data = 'This is my data';
  // Attach block to Tangle and log explorer link
  const sendResult = await sendData(client, tag, data);
  const blockId = sendResult.blockId;
  console.log(consoleColor, 'Attached block:');
  console.log(explorerURL + 'block/' + blockId, '\n');
  // Wait for block confirmation by milestone and read it with proof of inclusion from INX plugin
  const result = await getNotarization(client, nodeURL, blockId);
  // Store block with proof of inclusion in local json file
  if (result != false) {
    const filePath = `./notarized-block.json`;
    fs.writeFileSync(filePath, JSON.stringify(result, null, 4));
    console.log(consoleColor, 'Block successfully notarized and stored at:');
    console.log(filePath, '\n');
    console.log(
      consoleColor,
      'The Notarized block can now be handed over to the verifier',
    );
  }
}
// Function that regularly checks for block confirmation and returns proof of inclusion if confirmed after n tries
async function getNotarization(client, nodeURL, blockId) {
  try {
    console.log(
      consoleColor,
      'Wait for milestone confirmation to get notarized block:',
    );
    let i = 0;
    while (i < 10) {
      i++;
      // Function waits for a certain time between iterations
      await sleep(1000);
      const blockMetadata = await client.blockMetadata(blockId);
      // If a block was referenced by a milestone, the node will return its metadata with the key 'referencedByMilestoneIndex', otherwise the key won't be there
      if ('referencedByMilestoneIndex' in blockMetadata) {
        console.log(
          `Try ${i}: Block was referenced by milestone #${blockMetadata.referencedByMilestoneIndex}`,
          '\n',
        );
        // Call "create" endpoint of PoI plugin with blockId and return the result
        const poiPluginUrl = `${nodeURL}/api/poi/v1/create/${blockId}`;
        const response = await fetch(poiPluginUrl);
        const result = await response.json();
        return result;
      } else {
        console.log(`Try ${i}: Block was not yet referenced by a milestone`);
      }
    }
    console.log(`Block was not referenced by a milestone after ${i} seconds.`);
    return false;
  } catch (error) {
    console.log(error);
  }
}
function sleep(ms) {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}
run().catch((err) => console.error(err));
```

### Verify Notarization Script

Since you want to verify the notarization, you will also need to create a new file called `verify-notarization.js`. 

The `verify-notorization.js` you created will do the following:

1. Read the block with the notarization from your local folder.
2. Verify the block with the notarization.

You should create a new file called `verify-notarization.js` and add the following code:

```javascript
const { TransactionHelper } = require('@iota/iota.js');
const fs = require('fs');
const fetch = require('node-fetch')
// Network configuration
const { networkConfig } = require("./networkConfig.js");
const nodeURL = networkConfig.node;
const explorerURL = networkConfig.explorer;
// For the sake of this tutorial, some console output will be printed in a different color for better readability
const consoleColor = '\x1b[36m%s\x1b[0m';
async function run() {
    // Read and parse notarized block from file path
    const filePath = './notarized-block.json';
    const file = fs.readFileSync(filePath);
    const notarizedBlock = JSON.parse(file);
    console.log(consoleColor, 'Successfully imported notarized block from path:');
    console.log(filePath, '\n');
    // Generate blockId from block content and log explorer link
    // The blockId is defined as the BLAKE2b-256 hash of the entire serialized block
    const blockId = TransactionHelper.calculateBlockId(notarizedBlock.block);
    console.log(consoleColor, 'Notarized block:');
    console.log(explorerURL+"block/"+blockId, '\n');
    // Verify provided notarization/proof of inclusion for block
    const validity = await verifyNotarization(nodeURL, notarizedBlock);
    console.log(consoleColor, 'Validity of provided notarization:');
    console.log(validity, '\n');
}
async function verifyNotarization(nodeURL, notarizedBlock) {
    // Call "validate" endpoint of PoI plugin with notarized block and return boolean answer
    const poiPluginUrl = `${nodeURL}/api/poi/v1/validate`;
    const response = await fetch(poiPluginUrl, {
        method: 'POST', 
        body: JSON.stringify(notarizedBlock),
        headers: { 'Content-Type': 'application/json' }
    })
    const result = await response.json();
    return result.valid;
}
run().catch((err) => console.error(err));
```
````

> ### Run
> 
> Now you can execute the two created files in the correct order (First create-notarization then verify-notarization) and check the log output to follow along.
> 
> #### Create Notarization
> 
> ```bash
> node create-notarization.js
> ```
> 
> ##### Steps
> 
> 1. Setup client and define block content
> 2. Attach block to the Tangle
> 3. Wait for block to be confirmed by a milestone
> 4. Read block from Tangle together with notarization
> 5. Store block with notarization in a JSON file within your local folder
> 
> Your console output should look something like this:
> 
> ```
> Attached block:
> http://localhost:8082/dashboard/explorer/block/<id_of_your_notarized_block>
> 
> Wait for milestone confirmation to get notarized block:
> Try 1: Block was not yet referenced by a milestone
> Try 2: Block was not yet referenced by a milestone
> Try 3: Block was referenced by milestone <number_of_milestone_referencing_your_block>
> 
> Block successfully notarized and stored at:
> ./notarized-block.json
> 
> Notarized block can now be handed over to the verifier
> ```
> 
> #### Verify Notarization
> 
> ```bash
> node verify-notarization.js
> ```
> 
> ##### Steps
> 
> 1. Read block with notarization from your local folder
> 2. Verify block with notarization
> 
> Your console output should look something like this:
> 
> ```
> Successfully imported notarized block from path:
> ./notarized-block.json
> 
> Notarized block:
> http://localhost:8082/dashboard/explorer/block/<id_of_your_notarized_block>
> 
> Validity of provided notarization:
> true
> ```

Reviewer 2: ````suggestion
## Run Your Scripts

Once you have [created your scripts](#create-your-scripts), you can execute the two created files in the order. 
Naturally, you will need to run `create-notarization` before you can run `verify-notarization` and check the log output to follow along.

### Create the Notarization

You can create the notarization in the Tangle by running the following command:

```bash
node create-notarization.js
```

Your console output should look something like this:

```plaintext
Attached block:
http://localhost:8082/dashboard/explorer/block/<id_of_your_notarized_block>
Wait for milestone confirmation to get notarized block:
Try 1: Block was not yet referenced by a milestone
Try 2: Block was not yet referenced by a milestone
Try 3: Block was referenced by milestone <number_of_milestone_referencing_your_block>
Block successfully notarized and stored at:
./notarized-block.json
Notarized block can now be handed over to the verifier
```

### Verify Notarization

You can verify the notarization by running the following command:

```bash
node verify-notarization.js
```

Your console output should look something like this:

```plaintext
Successfully imported notarized block from path:
./notarized-block.json
Notarized block:
http://localhost:8082/dashboard/explorer/block/<id_of_your_notarized_block>
Validity of provided notarization:
true
```
````

# Tokenless Data Notarization

> - This tutorial shows how one can proof that a block was part of the past Tangle, even if this part of the Tangle was already pruned from network nodes

Reviewer 3: ```suggestion
- This tutorial shows how one can prove that a block was part of the past Tangle, even if this part of the Tangle was already pruned from network nodes
```

> - This tutorial shows how one can proof that a block was part of the past Tangle, even if this part of the Tangle was already pruned from network nodes

Reviewer 3: ```suggestion
- This tutorial shows how one can proof that a block was part of the past Tangle, even if this part of the Tangle was already pruned from network nodes.
```

## User Story

- A **Prover** wants to prove to a **Verifier**, that a dataset/file was not altered since a specific time in the past by notarizing it via the public and permissionless Tangle
- A good example could be the government (**Verifier**) obligating companies (**Prover**) to write their daily CO2 emissions to the Tangle

> - The government has no interest in providing storage for all the companies, but wants to receive veriably immutable data in the case of an audit, therefore all the data remains with the respective company until an audit occurs

Reviewer 3: ```suggestion
- The government has no interest in providing storage for all the companies but wants to receive verifiable, immutable data in the case of an audit therefore all the data remains with the respective company until an audit occurs
```

> - The government has no interest in providing storage for all the companies, but wants to receive veriably immutable data in the case of an audit, therefore all the data remains with the respective company until an audit occurs

Author: Done

### Architecture

- This pattern does not rely on a Permanode storing the full Tangle history
- The party writing to the Tangle (**Prover**) is the same party reading from it, which allows reading a block by its block Id
- All a **Verifier** needs to verify this proof, is the chain of milestones back to the milestone that referenced this exact block
  - Every milestone directly references the previous milestone
  - This creates a unique and verifiable chain of milestones in the Tangle
  - This means if the **Verifier's** notarization plugin didn't record the milestone history already, it could even download the milestone chain from another source and verify its correctness before using it for further notarization verifications
- Since the pattern assumes trusted access to the network, it is strongly recommended that both, the **Prover** and the **Verifier** operate their own network node and run the [INX Notarization/PoI Plugin](https://github.com/iotaledger/inx-poi)

> ![alt text](proof-inclusion-of-a-block-architecture.png)

Author: Done

> ### Used Components

Author: Done

    const blockId = TransactionHelper.calculateBlockId(notarizedBlock.block);
    console.log(consoleColor, 'Notarized block:');
    console.log(explorerURL+"block/"+blockId, '\n');

    // Verify provided notarization/proof of inclusion for block
    const validity = await verifyNotarization(nodeURL, notarizedBlock);
    console.log(consoleColor, 'Validity of provided notarization:');
    console.log(validity, '\n');
}

async function verifyNotarization(nodeURL, notarizedBlock) {
    // Call "validate" endpoint of PoI plugin with notarized block and return boolean answer
    const poiPluginUrl = `${nodeURL}/api/poi/v1/validate`;
    const response = await fetch(poiPluginUrl, {
        method: 'POST', 
        body: JSON.stringify(notarizedBlock),
        headers: { 'Content-Type': 'application/json' }
    })
    const result = await response.json();

    return result.valid;
}

run().catch((err) => console.error(err));
```

### Execution Steps

Now you can execute the two created files in the correct order (First create-notarization then verify-notarization) and check the log output to follow along.

#### Create Notarization

> _Command_

Author: Gone

```bash
node create-notarization.js
```

_Steps_

1. Setup client and define block content
2. Attach block to the Tangle
3. Wait for block to be confirmed by a milestone
4. Read block from Tangle together with notarization
5. Store block with notarization in a JSON file within your local folder

_Console output should look something like this_

```console
Attached block:
http://localhost:8082/dashboard/explorer/block/<id_of_your_notarized_block>

Wait for milestone confirmation to get notarized block:
Try 1: Block was not yet referenced by a milestone
Try 2: Block was not yet referenced by a milestone
Try 3: Block was referenced by milestone <number_of_milestone_referencing_your_block>

Block successfully notarized and stored at:
./notarized-block.json

Notarized block can now be handed over to the verifier
```

#### Verify Notarization

_Command_

```bash
node verify-notarization.js
```

_Steps_

1. Read block with notarization from your local folder
2. Verify block with notarization

> _Console output should look something like this_

Author: Done

> ```console

Author: Done

![PoI-Architecture](./images/proof-inclusion-of-a-block-architecture.png)

---

## Prerequisites

### Network

- A running [Hornet Private Tangle](https://github.com/iotaledger/hornet/tree/develop/private_tangle). You should use the **develop** branch to have the latest [Stardust](https://blog.shimmer.network/stardust-upgrade-in-a-nutshell/) features.
- The node should run the [INX-POI](https://github.com/iotaledger/inx-poi). The node(s) will start the INX-POI plugin by default.

### Development Environment and Libraries

- [Node.js](https://nodejs.org/en/).

## Set Up

### Prepare Your Development Environment

1. Create a new folder for the tutorial in a location of your choice, for example:

```bash
mkdir notarization-tutorial
cd notarization-tutorial
```

2.  You should run the Node.js initializer to configure the project.

```bash
npm init --yes
```

3. Add [`@iota/iota.js`](https://www.npmjs.com/package/@iota/iota.js) as a dependency by running the following command:

```bash
npm install @iota/iota.js
```

---

## Create Your Scripts

> ### Network Configuration Script

Reviewer: ```suggestion
### Network Configuration Script

```

  console.log(explorerURL + 'block/' + blockId, '\n');
```

Wait for block confirmation by a milestone and read the block with proof of inclusion from INX plugin. The function `getNotarization()` will be explained in more detail in the next section.
```javascript
  // Wait for block confirmation by milestone and read it with proof of inclusion from INX plugin
  const result = await getNotarization(client, nodeURL, blockId);
```

This part will only be exectured, if the respective block was confirmed by a milestone after a defined time. If that is the case, the returned notarization result will be stored in a *.json file and the file path will be logged out.
```javascript
  // Store block with proof of inclusion in local json file
  if (result != false) {
    const filePath = `./notarized-block.json`;
    fs.writeFileSync(filePath, JSON.stringify(result, null, 4));

    console.log(
      consoleColor,
      'Block successfully notarized and stored at:'
    );
    console.log(filePath, '\n');

    console.log(
      consoleColor,
      'The Notarized block can now be handed over to the verifier',
    );
  }
}
```

#### 3. Get notarization

> As described in the previous step, the `getNotarization()` function is called from within `main()`. The function fetches the metadata of a given blockId 10 times while pausing for 1 second after each try via the `sleep()` function. As soon as a block was referenced by a milestone, the notarization for the block is fetched from the proof of inclusion plugin and  returned by the function. In case the block was not referenced by a milestone after 10 seconds, the function returns `false`.

Reviewer: ```suggestion
As described in the previous step, the `getNotarization()` function is called from within `main()`. As soon as a block was referenced by a milestone, the notarization for the block is fetched from the proof of inclusion plugin and returned by the function. The function tries to fetch the metadata of a given blockId 10 times, waiting for a second before each try. In case the block was still not referenced by a milestone, the function returns `false`.
```

I brought the more important information forward and rephrased it a bit.

  try {
    console.log(
      consoleColor,
      'Wait for milestone confirmation to get notarized block:',
    );

    let i = 0;
    while (i < 10) {
      i++;
      // Function waits for a certain time between iterations
      await sleep(1000);

      const blockMetadata = await client.blockMetadata(blockId);

      // If a block was referenced by a milestone, the metadata contains the field 'referencedByMilestoneIndex'
      if ('referencedByMilestoneIndex' in blockMetadata) {
        console.log(
          `Try ${i}: Block was referenced by milestone #${blockMetadata.referencedByMilestoneIndex}`,
          '\n',
        );

        // Call "create" endpoint of PoI plugin with blockId and return the result
        const poiPluginUrl = `${nodeURL}/api/poi/v1/create/${blockId}`;
        const response = await fetch(poiPluginUrl);
        const result = await response.json();

        return result;

      } else {
        console.log(`Try ${i}: Block was not yet referenced by a milestone`);
      }
    }
    console.log(`Block was not referenced by a milestone after ${i} seconds.`);

    return false;

  } catch (error) {
    console.log(error);
  }
}
```

> #### 4. Delay function

Reviewer: ```suggestion
#### 4. Delay function

```

The function `sleep()` receives a number of milliseconds and acts as a time delay for whichever function is calling it.

```javascript
function sleep(ms) {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}
```

> #### 5. Execute main function

Reviewer: ```suggestion
#### 5. Execute main function

```

As a final step the function `main()` is called in order to trigger the process as described above.

```javascript
main().catch((err) => console.error(err));
```

---

### Verify Notarization Script

> Create a new file named `verify-notarization.js`. As with the previous script (`create-notarization.js`), the code below is separated into distinct code snippets together with an explanation. After you have followed along the different steps, you can copy all snippets into the newly created file, following their respective order.

Reviewer: ```suggestion
Create a new file named `verify-notarization.js`.

Just like with the [Create Notarization](#create-notarization-script) script, we broke the code into separate snippets. To make it work, copy all code snippets one after another into the file that you have just created.
```

Similar to the previous suggestion. I dropped "to help you understand it better" to make space for the reference to the previous script. This bit is not crucial and the reader will probably recall it anyway.

> #### 1. Imports and parameters
> This part imports all necessary packages and network configuration parameters.

Reviewer: ```suggestion
#### 1. Imports and parameters

This part imports all necessary packages and network configuration parameters.
```

```javascript
const { TransactionHelper } = require('@iota/iota.js');
const fs = require('fs');
const fetch = require('node-fetch')

// Network configuration
const { networkConfig } = require("./networkConfig.js");
const nodeURL = networkConfig.node;
const explorerURL = networkConfig.explorer;

// For better readability, some console output will be printed in a different color
const consoleColor = '\x1b[36m%s\x1b[0m';
```

> #### 2. Main function
> The `main()` function will be called when running the `verify-notarization.js` file and consists of the following steps.

Reviewer: ```suggestion
#### 2. Main function

The `main()` function will be called when running the `verify-notarization.js` file and consists of the following steps:
```

> Declare the function, read the notarization file and parse it.
> ```javascript

Reviewer: ```suggestion
Declare the function, read the notarization file and parse it.

```javascript
```

async function run() {
    // Read and parse notarized block from file path
    const filePath = './notarized-block.json';
    const file = fs.readFileSync(filePath);
    const notarizedBlock = JSON.parse(file);

    console.log(consoleColor, 'Successfully imported notarized block from path:');
    console.log(filePath, '\n');
```

> Derive the blockId from the block content and log out the explorer link.

Reviewer: ```suggestion
Derive the blockId from the block content and log out the explorer link.

```

```javascript
    // Generate blockId from block content and log explorer link
    // The blockId is defined as the BLAKE2b-256 hash of the entire serialized block
    const blockId = TransactionHelper.calculateBlockId(notarizedBlock.block);
    console.log(consoleColor, 'Notarized block:');
    console.log(explorerURL+"block/"+blockId, '\n');
```

> Fetch validity (`true`/`false`) of the notarization from the proof of inclusion plugin. The function `verifyNotarization()` will be explained in more detail in the next section.

Reviewer: ```suggestion
Fetch validity (`true`/`false`) of the notarization from the proof of inclusion plugin. The function `verifyNotarization()` will be explained in more detail in the next section.

```

```javascript
    // Verify provided notarization/proof of inclusion for block
    const validity = await verifyNotarization(nodeURL, notarizedBlock);
    console.log(consoleColor, 'Validity of provided notarization:');
    console.log(validity, '\n');
}
```

> #### 3. Verify notarization

Reviewer: ```suggestion
#### 3. Verify notarization

```

As described in the previous step, the `verifyNotarization()` function is called from within `main()`. The function sends the notarized block to the `validate` endpoint of the proof of inclusion plugin and returns the booolean result.

```javascript
async function verifyNotarization(nodeURL, notarizedBlock) {
    // Call "validate" endpoint of PoI plugin with notarized block and return boolean answer
    const poiPluginUrl = `${nodeURL}/api/poi/v1/validate`;
    const response = await fetch(poiPluginUrl, {
        method: 'POST', 
        body: JSON.stringify(notarizedBlock),
        headers: { 'Content-Type': 'application/json' }
    })
    const result = await response.json();

    return result.valid;
}
```

> #### 4. Execute main function

Reviewer: ```suggestion
#### 4. Execute main function

```

> # Tokenless Data Notarization

Reviewer: Btw, is metadata left out intentionally or did we just forgot to add it?

@Dr-Electron

Reviewer 3: I guess we just forgot it. Not sure if @adrian-grassl even knows metadata 🤔

Reviewer 3: I think we can update this in a new PR if needed
